symforce/slam/imu_preintegration/imu_factor.h

```diff
@@ -0,0 +1,192 @@
+ *                                      {Var::GRAVITY},
+ *                                      {Var::EPSILON}}));
+ *     }
+ *
+ *     AppendOtherFactors(factors);
+ *
+ *     sym::Optimizerd optimizer(sym::DefaultOptimizerParams(), factors);
+ *
+ *     // Build Values
+ *     sym::Valuesd values;
+ *     for (int i = 0; i < key_frame_times.size(); i++) {
+ *       values.Set({Var::POSE, i}, sym::Pose3d());
+ *       values.Set({Var::VELOCITY, i}, Eigen::Vector3d::Zero());
+ *     }
+ *     for (int i = 0; i < key_frame_times.size() - 1; i++) {
+ *       values.Set({Var::ACCEL_BIAS, i}, GetAccelBiasEstimate(key_frame_times[i]));
+ *       values.Set({Var::GYRO_BIAS, i}, GetGyroBiasEstimate(key_frame_times[i]));
+ *     }
+ *     // gravity should point towards the direction of acceleration
+ *     values.Set({Var::GRAVITY}, (Eigen::Vector3d() << 0.0, 0.0, -9.8).finished());
+ *     values.Set({Var::EPSILON}, sym::kDefaultEpsilond);
+ *     // Initialize any other items in values ...
+ *
+ *     optimizer.Optimize(&values);
+ *   }
+ */
+template <typename Scalar>
+class ImuFactor {
+ private:
+  PreintegratedImuMeasurements<Scalar> preintegrated_measurements_;
+  Eigen::Matrix<Scalar, 9, 9> sqrt_info_;
+
+ public:
+  /**
+   * Construct an ImuFactor connecting two states from the (preintegrated) imu measurements
+   * between them.
+   */
+  ImuFactor(const ImuPreintegrator<Scalar>& preintegrator);
+
+  /**
+   * Construct a Factor object that can be passed to an Optimizer object given the keys to
+   * be passed to the function.
+   *
+   * Convenience method to avoid manually specifying which arguments are optimized.
+   */
+  sym::Factor<Scalar> Factor(const std::vector<Key>& keys_to_func);
+
+  /**
+   * Calculate a between factor on the product manifold of the pose and velocity where the prior
+   * is calculated from the preintegrated IMU measurements.
+   *
+   * Time step i is the time of the first IMU measurement of the interval.
+   * Time step j is the time after the last IMU measurement of the interval.
+   *
+   * Args:
+   *     pose_i: Pose at time step i (maps body frame at time i to world frame)
+   *     vel_i: Velocity at time step i (world frame)
+   *     pose_j: Pose at time step j (maps body frame at time j to world frame)
+   *     vel_j: Velocity at time step j (world frame)
+   *     accel_bias_i: The bias of the accelerometer measurements between timesteps i and j
+   *     gyro_bias_i: The bias of the gyroscope measurements between timesteps i and j
+   *     gravity: Acceleration due to gravity (in the same frame as pose_x and vel_x),
+   *              i.e., the vector which when subtracted from the accelerometer measurements
+   *              gives the true acceleration (up to bias and noise) of the IMU.
+   *     epsilon: epsilon used for numerical stability
+   *
+   * Outputs:
+   *     res: The 9dof local coordinate difference between predicted and estimated state
```

Comment:
It's not exactly the local coordinate difference right, it's also whitened by the covariance?

Reply:
Oh yeah. That's right. I'll say the "9dof whitened local coordinate difference..."